cedar-policy-validator/src/err.rs

```diff
@@ -112,204 +105,333 @@ impl HumanSyntaxParseError {
     ))]
-    ActionAttributesContainEmptySet(EntityUID),
-    /// An action entity (transitively) has an attribute of unsupported type (`ExprEscape`, `EntityEscape` or `ExtnEscape`).
-    /// This error variant should only be used when `PermitAttributes` is enabled.
+    pub struct ActionAttributesContainEmptySetError(pub(crate) EntityUID);
+
+    /// Unsupported action attribute error
+    #[derive(Debug, Diagnostic, Error)]
     #[error("action `{0}` has an attribute with unsupported JSON representation: {1}")]
-    UnsupportedActionAttribute(EntityUID, String),
-    /// Error when evaluating an action attribute
-    #[error(transparent)]
-    #[diagnostic(transparent)]
-    ActionAttrEval(EntityAttrEvaluationError),
-    /// Error thrown when the schema contains the `__expr` escape.
-    /// Support for this escape form has been dropped.
+    pub struct UnsupportedActionAttributeError(pub(crate) EntityUID, pub(crate) SmolStr);
+
+    /// Unsupported `__expr` escape error
+    #[derive(Debug, Clone, Diagnostic, Error)]
     #[error("the `__expr` escape is no longer supported")]
     #[diagnostic(help("to create an entity reference, use `__entity`; to create an extension value, use `__extn`; and for all other values, use JSON directly"))]
-    ExprEscapeUsed,
-    /// The schema used an extension type that the validator doesn't know about.
+    pub struct ExprEscapeUsedError {}
+
+    /// Action attribute evaluation error
+    #[derive(Debug, Diagnostic, Error)]
     #[error(transparent)]
     #[diagnostic(transparent)]
-    UnknownExtensionType(UnknownExtensionType),
-}
+    pub struct ActionAttrEvalError(#[from] pub(crate) EntityAttrEvaluationError);
 
-impl From<transitive_closure::TcError<EntityUID>> for SchemaError {
-    fn from(e: transitive_closure::TcError<EntityUID>) -> Self {
-        // we use code in transitive_closure to check for cycles in the action
-        // hierarchy, but in case of an error we want to report the more descriptive
-        // CycleInActionHierarchy instead of ActionTransitiveClosureError
-        match e {
-            transitive_closure::TcError::MissingTcEdge { .. } => {
-                SchemaError::ActionTransitiveClosure(Box::new(e))
-            }
-            transitive_closure::TcError::HasCycle(err) => {
-                SchemaError::CycleInActionHierarchy(err.vertex_with_loop().clone())
+    /// Unsupported feature error
+    #[derive(Debug, Diagnostic, Error)]
+    #[error("unsupported feature used in schema: {0}")]
+    #[diagnostic(transparent)]
+    pub struct UnsupportedFeatureError(#[from] pub(crate) UnsupportedFeature);
+
+    #[derive(Debug, Diagnostic, Error)]
+    pub enum SchemaError {
+        /// Error thrown by the `serde_json` crate during serialization
+        #[error(transparent)]
+        #[diagnostic(transparent)]
+        // no #[from], because if you just have a serde_json::Error you should choose between JsonDeserialization and JsonSerialization appropriately
```

Comment:
Should the `JsonDeserialization` variant also not have `#[from]`?

Reply:
Good catch. Fixed.